### PR TITLE
log: fix stack overflow when flushing large log lines

### PR DESCRIPTION
**Pull request by ifed01**
See http://tracker.ceph.com/issues/14707

Signed-off-by: Igor Fedotov <ifedotov@mirantis.com>

### CODE REVIEW

File: src/log/Log.cc [hunk 1]
```diff
@@ -277,17 +277,26 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
     e->hint_size();
     if (do_fd || do_syslog || do_stderr) {
       size_t buflen = 0;
-      char buf[80 + e->size()];
+
+      char * buf;
+      size_t buf_size = 80 + e->size();
+      bool need_dynamic = buf_size >= 0x10000; //avoids >64K buffers allocation at stack
+      char buf0[ need_dynamic ? buf_size : 1];
+      if( need_dynamic ) {
```
Comment (liewegas) [inline src/log/Log.cc]: if (need_dynamic) {

File: src/log/Log.cc [hunk 2]
```diff
@@ -277,17 +277,26 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
     e->hint_size();
     if (do_fd || do_syslog || do_stderr) {
       size_t buflen = 0;
-      char buf[80 + e->size()];
+
+      char * buf;
+      size_t buf_size = 80 + e->size();
+      bool need_dynamic = buf_size >= 0x10000; //avoids >64K buffers allocation at stack
+      char buf0[ need_dynamic ? buf_size : 1];
```
Comment (liewegas) [inline src/log/Log.cc]: no space inside brackets

File: src/log/Log.cc [hunk 3]
```diff
@@ -277,17 +277,26 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
     e->hint_size();
     if (do_fd || do_syslog || do_stderr) {
       size_t buflen = 0;
-      char buf[80 + e->size()];
+
+      char * buf;
```
Comment (liewegas) [inline src/log/Log.cc]: char *buf

File: src/log/Log.cc [hunk 4]
```diff
@@ -304,7 +313,8 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
         if (r < 0)
           cerr << "problem writing to " << m_log_file << ": " << cpp_strerror(r) << std::endl;
       }
-
+      if( need_dynamic )
```
Comment (liewegas) [inline src/log/Log.cc]: if (need_dynamic)

**Comment by liewegas** [comment]
Do you mind adding a test case to unittest_log for this?

Thanks!

**Comment by ifed01** [comment]
@liewegas - done

**Comment by liewegas** [comment]
Thanks!

**Pull request merged**